compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +447,294 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+  // OpOperand of "another_op" will be replaced:
+  //
+  // %0 = "some_op"() : () -> (tensor<?xf32>)
+  // %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+  //   %0_clone = "some_op"() : () -> (tensor<?xf32>)
+  //   %1 = "another_op"(%0_clone) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //   flow.return %1 : tensor<?xf32>
+  // }
+  // %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //
+  // Note: If it weren't for "yet_another_use", the original "some_op" could
+  // fold away.
+  //
+  // Note: If there were no "another_op" in the dispatch region, the IR would
+  // not be modified and this function would fail.
+  if (domInfo.properlyDominates(target, regionOp)) {
+    SmallVector<OpOperand *> consumers;
+    for (OpOperand &use : target->getUses())
+      if (regionOp->isProperAncestor(use.getOwner())) consumers.push_back(&use);
+    // We currently only fuse the consumers inside the regionOp. If there are
+    // no consumers, the op will not be moved inside the regionOp. This could
+    // be customized with a flag, such that target is always cloned into the
+    // regionOp and uses that post-dominate the regionOp are updated to the a
+    // newly added regionOp result.
+    // E.g., in the above example, flow.return would also return %0_clone and
+    // "yet_another_use" would use %r#1 instead of %0.
+    if (!consumers.empty()) {
+      rewriter.setInsertionPointToStart(&body);
+      Operation *cloned = rewriter.clone(*target);
+      for (OpOperand *consumer : consumers) {
+        rewriter.updateRootInPlace(consumer->getOwner(), [&]() {
+          consumer->set(cloned->getResult(
+              consumer->get().cast<OpResult>().getResultNumber()));
+        });
+      }
+
+      // Delete op if it no longer has any uses.
+      if (target->use_empty()) rewriter.eraseOp(target);
+
+      return success();
+    }
+  }
+
+  // Case 2: `target` post-dominates `regionOp`. The target is cloned into the
+  // regionOp and all original uses of the target are replaced with a newly
+  // added regionOp result.
+  //
+  // Example:
+  //
+  // %r = flow.dispatch.region -> (tensor<?xf32>{%d}) {
+  //   %0 = ...
+  //   flow.return %0 : tensor<?xf32>
+  // }
+  // %1 = "some_op"(%r) : (tensor<?xf32>) -> (tensor<?xf32>)
+  // "some_use"(%1) : (tensor<?xf32>) -> ()
+  //
+  // In this example, "some_op" will be cloned into the dispatch region and all
+  // of its uses are replaced.
+  //
+  // %r:2 = flow.dispatch.region -> (tensor<?xf32>{%d}, tensor<?xf32>{%e}) {
+  //   %0 = ...
+  //   %1_clone = "some_op"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+  //   flow.return %0, %1_clone : tensor<?xf32>, tensor<?xf32>
+  // }
+  // "some_use"(%r#1) : (tensor<?xf32>) -> ()
+  //
+  // Note: If there is no use of %r#0, the dispatch region can canonicalize to
+  // an op with only one result.
+  if (!domInfo.properlyDominates(regionOp, target)) return failure();
```

Comment:
Do you really need dominator information here. You can just use the SSA producer-consumer relationship.... 
Also not convinced this is needed. It is better/easier to create the region with the consumer in this case...

Reply:
Dropped the DominatorInfo usage. It is not needed.

Reply:
It is still used in an `assert` for easier debbuging